src/alarm_heap.cpp

```diff
@@ -0,0 +1,326 @@
+  int rc = pthread_create(&_heap_sender_thread, NULL, heap_sender_function, this);
+
+  if (rc < 0)
+  {
+    // LCOV_EXCL_START
+    printf("Failed to start heap pop thread: %s", strerror(errno));
+    exit(2);
+    // LCOV_EXCL_STOP
+  }
+}
+
+AlarmHeap::~AlarmHeap()
+{
+  pthread_mutex_lock(&_lock);
+  _terminated = true;
+  _cond->signal();
+  pthread_mutex_unlock(&_lock);
+
+  pthread_join(_heap_sender_thread, NULL);
+  delete _cond; _cond = NULL;
+
+  pthread_mutex_destroy(&_lock);
+
+  delete _alarm_trap_sender; _alarm_trap_sender = NULL;
+  _alarm_heap.clear();
+
+  for (AllAlarmsStateMap::iterator it = _all_alarms_state.begin();
+                               it != _all_alarms_state.end();
+                               it++)
+  {
+    delete (*it).second;
+  }
+}
+
+void* AlarmHeap::heap_sender_function(void* data)
+{
+  ((AlarmHeap*)data)->heap_sender();
+  return NULL;
+}
+
+void AlarmHeap::heap_sender()
+{
+  pthread_mutex_lock(&_lock);
+
+  while (!_terminated)
+  {
+    uint64_t time_now_in_ms = Utils::get_time();
+    AlarmInHeap* alarm_in_heap = (AlarmInHeap*)_alarm_heap.get_next_timer();
+
+    while ((alarm_in_heap) &&
+           (alarm_in_heap->get_pop_time() <= time_now_in_ms))
+    {
+      // While the next alarm in the heap is due to be sent, we:
+      //  - Pull out the alarm definition for its index and severity.
+      //  - Pass the alarm definition to the trap sender (which is responsible
+      //    for actually sending any INFORMs).
+      //  - Update the master view of the current severity for this alarm (by
+      //    updating the _all_alarms_state map).
+      //  - Finally, we then remove the alarm from the heap.
+      AlarmTableDef& alarm_table_def =
+              _alarm_table_defs->get_definition(alarm_in_heap->index(),
+                                                alarm_in_heap->severity());
+
+      if (alarm_table_def.is_valid())
+      {
+        _alarm_trap_sender->send_trap(alarm_table_def);
+        AllAlarmsStateMap::iterator alarm = _all_alarms_state.find(alarm_in_heap->index());
+
+        if (alarm != _all_alarms_state.end())
+        {
+          (*alarm).second->set_severity(alarm_table_def);
+        }
+      }
+
+      alarm_in_heap->remove_from_heap();
+      alarm_in_heap = (AlarmInHeap*)_alarm_heap.get_next_timer();
+    }
+
+    // The next alarm in the heap isn't due to pop yet. Wait until it's due.
+    if (alarm_in_heap)
+    {
+      // The next alarm in the heap isn't due to pop yet. Wait until it's due.
+      struct timespec time;
+      time.tv_sec = alarm_in_heap->get_pop_time() / 1000;
+      time.tv_nsec = (alarm_in_heap->get_pop_time() % 1000) * 1000000;
+      _cond->timedwait(&time);
+    }
+    else
+    {
+      // There's no alarms in the heap. Wait until we're signalled.
+      _cond->wait();
+    }
+  }
+
+  pthread_mutex_unlock(&_lock);
+}
+
+void AlarmHeap::update_alarm_in_heap(AlarmInfo* alarm_info,
```

Comment:
I wonder if this could have a name that describes what it's doing more conceptually (`schedule_severity_change`)?